### PR TITLE
Convert Terser to optimizer

### PR DESCRIPTION
**Pull request by wbinnssmith**
This converts the terser transformer to an optimizer. This:
* Optimizes the prelude, which never went through the transformer
* Upgrades terser to 4.0
* Continues to honor the user's uglify/terser config
* Uses terser's default for mangling (top-level and eval are false) in the fallback config

Test Plan: `yarn test`, and use the kitchen sink example with scope hoisting on and off to verify executable output. Verify in the non-scopehoisted case that `parcelRequire` isn't mangled.

### CODE REVIEW

**Comment by mischnic** [comment]
Previous discussion on whether terser should run on Asset or Packager level: https://github.com/parcel-bundler/parcel/issues/2310

This is *much* better for code size, but might have longer built time (and memory usage?).

**Comment by wbinnssmith** [comment]
Re: build time, given terser (and most optimizers in general) are likely only enabled with `minify: true` which isn't the default at development, I'm not really concerned unless it's really significant.

**Comment by mischnic** [comment]
> top-level false

Could we enable this?
> set to true if you wish to enable top level variable and function name mangling and **to drop unused variables and functions.**

**Comment by wbinnssmith** [comment]
> Could we enable this?

I think we can, given Parcel controls the top-level. Is it possible that after scope hoisting and tree shaking there are unused variables?

**Comment by devongovett** [comment]
I think that option is for mangling, not dead code elimination

File: packages/optimizers/terser/src/TerserOptimizer.js [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+export default new Optimizer({
+  async optimize({contents, bundle, options}) {
+    if (!options.minify) {
+      return contents;
+    }
+
+    if (typeof contents !== 'string') {
+      throw new Error(
+        'TerserOptimizer: Only string contents are currently supported'
```
Comment (devongovett) [inline packages/optimizers/terser/src/TerserOptimizer.js]: ahh this is interesting. maybe once we introduce ast sharing for optimizers, we could end up with a `getCode`, `getBuffer`, `getAST` etc. on a `Bundle` similar to the ones in `Asset`

**Pull request merged**